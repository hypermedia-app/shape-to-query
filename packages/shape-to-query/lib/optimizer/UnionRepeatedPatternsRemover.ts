import Processor from '@hydrofoil/sparql-processor'
import type {
  BgpPattern, Expression,
  GroupPattern,
  Pattern,
  Triple,
  UnionPattern,
  ValuePatternRow,
  ValuesPattern,
} from 'sparqljs'
import { tripleEquals } from '@hydrofoil/sparql-processor/util/triple.js'
import { valuesHasRow } from '@hydrofoil/sparql-processor/util/values.js'
import { match, P } from 'ts-pattern'
import type { DataFactory, Variable } from '@rdfjs/types'

/**
 * Removes patterns from a union that are already present in the outer scope
 */
export class UnionRepeatedPatternsRemover extends Processor {
  declare patterns: Triple[][]
  declare values: ValuesPattern[][]
  declare union: UnionPattern
  declare group: GroupPattern

  constructor(factory: DataFactory) {
    super(factory)
    this.patterns = []
    this.values = []
  }

  processPatterns(patterns: Pattern[]): Pattern[] {
    if (this.union) {
      const processed = super.processPatterns(patterns)
      return processed.filter(removeEmpty)
    }

    this.patterns.push(patterns.flatMap(pattern => {
      return 'triples' in pattern ? pattern.triples : []
    }))
    this.values.push(patterns.filter((p): p is ValuesPattern => p.type === 'values'))

    const processed = super.processPatterns(patterns).filter(removeEmpty)

    this.values.pop()
    this.patterns.pop()

    return processed
  }

  processUnion(union: UnionPattern): Pattern {
    this.union = union

    const processed = super.processUnion(union)

    this.union = undefined
    if (processed.type !== 'union') {
      return processed
    }

    if (processed.patterns.length === 1) {
      if (processed.patterns[0].type === 'group') {
        return {
          type: 'optional',
          patterns: processed.patterns[0].patterns,
        }
      }

      return {
        type: 'optional',
        patterns: processed.patterns,
      }
    }

    return processed
  }

  processValues(values: ValuesPattern): Pattern {
    if (this.union) {
      values = {
        type: 'values',
        values: values.values.filter(row => {
          return !this.values.some(values => values.some(valuesHasRow(row)) && !this.rowOrTripleUsedInBind(row))
        }),
      }
    }

    return super.processValues(values)
  }

  processGroup(group: GroupPattern): Pattern {
    this.group = group
    const processed = super.processGroup(group)
    this.group = undefined
    return processed
  }

  processBgp(bgp: BgpPattern): BgpPattern {
    if (this.union) {
      const allTriplesInUpperScope = bgp.triples.every(triple => {
        return this.patterns.some(triples => triples.some(tripleEquals(triple)) && !this.rowOrTripleUsedInBind(triple))
      })

      if (allTriplesInUpperScope) {
        // remove all triples from the bgp
        bgp.triples = []
      }
    }

    return super.processBgp(bgp)
  }

  private rowOrTripleUsedInBind(row: ValuePatternRow | Triple): boolean {
    if (!this.group) {
      return false
    }

    const rowValues = Object.values(row)
    const _valuesUsedInExpression = valuesUsedInExpression(row)

    return this.group.patterns.some(pattern => {
      if (pattern.type !== 'bind') {
        return false
      }

      return match(pattern)
        .with({ expression: P.union({ type: 'operation' }, { type: 'functionCall' }) }, ({ expression }) => {
          return expression.args.some(_valuesUsedInExpression)
        })
        .with({ expression: { type: 'aggregate' } }, ({ expression }) => {
          return !('termType' in expression.expression) && _valuesUsedInExpression(expression.expression)
        })
        .with({ expression: P.array() }, ({ expression }) => expression.some(_valuesUsedInExpression))
        .with({ expression: { termType: 'Variable' } }, ({ expression }) => rowValues.some(expression.equals.bind(expression)))
        .with({ expression: { termType: P.any } }, () => false)
        .exhaustive()
    })
  }
}

function valuesUsedInExpression(row: ValuePatternRow | Triple) {
  let variableIsUsed: (variable: Variable) => boolean
  if ('subject' in row) {
    variableIsUsed = variable => variable.equals(row.subject) ||
      ('termType' in row.predicate && variable.equals(row.predicate)) ||
      variable.equals(row.object)
  } else {
    variableIsUsed = variable => row[`?${variable.value}`] !== undefined
  }

  return function matcher(expression: Expression) {
    /* eslint-disable @typescript-eslint/ban-ts-comment */
    // @ts-ignore TS2589: Type instantiation is excessively deep and possibly infinite.
    return match(expression)
      .with(P.union({ type: 'operation' }, { type: 'functionCall' }), ({ args }) => {
        // @ts-ignore TS2589: Type instantiation is excessively deep and possibly infinite.
        return args.some(matcher)
      })
      .with({ type: 'aggregate' }, ({ expression }) => {
        return !('termType' in expression) && matcher(expression)
      })
      .with(P.array(), (tuple) => tuple.some(matcher))
      .with({ termType: 'Variable' }, variableIsUsed)
      .with({ termType: P.any }, () => false)
      .exhaustive()
  }
}

function removeEmpty(pattern: Pattern): boolean {
  switch (pattern.type) {
    case 'values':
      return pattern.values.length > 0
    case 'group':
      return pattern.patterns.length > 0
    case 'bgp':
      return pattern.triples.length > 0
  }

  return true
}
