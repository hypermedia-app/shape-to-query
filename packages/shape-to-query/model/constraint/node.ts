import { sh } from '@tpluscode/rdf-ns-builders'
import type sparqljs from 'sparqljs'
import type { ModelFactory } from '../ModelFactory.js'
import type { NodeShape } from '../NodeShape.js'
import type { Parameters, PropertyShape } from './ConstraintComponent.js'
import ConstraintComponent, { assertTerm } from './ConstraintComponent.js'

export class NodeConstraintComponent extends ConstraintComponent {
  constructor(public readonly shape: NodeShape) {
    super(sh.NodeConstraintComponent)
  }

  static * fromShape(shape: PropertyShape, factory: ModelFactory) {
    const nodes = shape.get(sh.node) || []
    for (const node of nodes) {
      assertTerm(node)
      yield new NodeConstraintComponent(factory.nodeShape(node.pointer))
    }
  }

  buildPatterns({ valueNode, variable, ...arg }: Parameters): [sparqljs.GroupPattern] | sparqljs.Pattern[] {
    const patterns = this.shape.buildConstraints({
      ...arg,
      variable,
      focusNode: valueNode,
      valueNode: variable(),
    })

    if (!patterns.length) {
      return patterns
    }

    return [{
      type: 'group',
      patterns,
    }]
  }
}
