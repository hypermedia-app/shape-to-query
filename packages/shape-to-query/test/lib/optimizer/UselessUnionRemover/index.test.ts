import * as fs from 'node:fs'
import * as path from 'node:path'
import chai, { expect } from 'chai'
import $rdf from '@zazuko/env'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import sparqljs from 'sparqljs'
import { globSync } from 'glob'
import { UnionRepeatedPatternsRemover } from '../../../../lib/optimizer/UnionRepeatedPatternsRemover.js'

const __dirname = path.dirname(new URL(import.meta.url).pathname)

describe('UselessUnionRemover', () => {
  chai.use(jestSnapshotPlugin())

  const parser = new sparqljs.Parser()
  const generator = new sparqljs.Generator()
  const processor = new UnionRepeatedPatternsRemover($rdf)

  function parseQuery(fileName: string) {
    return parser.parse(fs.readFileSync(path.resolve(__dirname, fileName)).toString())
  }

  function testCase(filePath: string) {
    const fileName = path.basename(filePath)

    it(fileName, () => {
      const query = parseQuery(filePath)
      const result = processor.process(query)
      expect(generator.stringify(result)).toMatchSnapshot()
    })
  }

  globSync(path.resolve(__dirname, '*.rq')).forEach(testCase)
})
