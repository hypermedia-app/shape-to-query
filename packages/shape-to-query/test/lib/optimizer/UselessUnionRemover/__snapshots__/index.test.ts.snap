// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UselessUnionRemover multiple-patterns.rq 1`] = `
"PREFIX schema: <http://schema.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
CONSTRUCT {
  ?resource8 schema:containedInPlace ?resource1.
  ?resource8 schema:identifier ?resource11.
  ?resource8 schema:name ?resource12.
  ?resource8 schema:description ?resource13.
  ?resource8 schema:position ?resource14.
  ?resource8 schema:alternateName ?resource15.
  ?resource8 rdf:type ?resource10.
}
WHERE {
  VALUES ?resource1 {
    <https://ld.admin.ch/country/CHE>
  }
  ?resource1 ^schema:containedInPlace ?resource8.
  ?resource8 rdf:type ?resource10.
  VALUES ?resource10 {
    <https://environment.ld.admin.ch/foen/nfi/UnitOfReference/366>
  }
  { ?resource8 schema:identifier ?resource11. }
  UNION
  { ?resource8 schema:name ?resource12. }
  UNION
  { ?resource8 schema:description ?resource13. }
  UNION
  { ?resource8 schema:position ?resource14. }
  UNION
  { ?resource8 schema:alternateName ?resource15. }
}"
`;

exports[`UselessUnionRemover partial-match.rq 1`] = `
"PREFIX schema: <http://schema.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
CONSTRUCT {
  ?resource8 schema:containedInPlace ?resource1.
  ?resource8 schema:identifier ?resource11.
  ?resource8 schema:name ?resource12.
  ?resource8 schema:description ?resource13.
  ?resource8 schema:position ?resource14.
  ?resource8 schema:alternateName ?resource15.
  ?resource8 rdf:type ?resource10.
}
WHERE {
  VALUES ?resource1 {
    <https://ld.admin.ch/country/CHE>
  }
  ?resource1 ^schema:containedInPlace ?resource8.
  ?resource8 rdf:type ?resource10.
  VALUES ?resource10 {
    <https://environment.ld.admin.ch/foen/nfi/UnitOfReference/366>
  }
  { ?resource8 schema:identifier ?resource11. }
  UNION
  { ?resource8 schema:name ?resource12. }
  UNION
  { ?resource8 schema:description ?resource13. }
  UNION
  { ?resource8 schema:position ?resource14. }
  UNION
  { ?resource8 schema:alternateName ?resource15. }
  UNION
  {
    ?resource8 rdf:type ?resource10.
    ?resource10 (rdfs:subClassOf*) schema:Place.
  }
}"
`;

exports[`UselessUnionRemover remove-useless-union.rq 1`] = `
"PREFIX schema: <http://schema.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
CONSTRUCT {
  ?resource8 schema:containedInPlace ?resource1.
  ?resource8 schema:identifier ?resource11.
  ?resource8 rdf:type ?resource10.
}
WHERE {
  VALUES ?resource1 {
    <https://ld.admin.ch/country/CHE>
  }
  ?resource1 ^schema:containedInPlace ?resource8.
  ?resource8 rdf:type ?resource10.
  VALUES ?resource10 {
    <https://environment.ld.admin.ch/foen/nfi/UnitOfReference/366>
  }
  ?resource8 schema:identifier ?resource11.
  FILTER(REGEX(?resource11, \\"^test\\"))
  
}"
`;

exports[`UselessUnionRemover single-pattern.rq 1`] = `
"PREFIX schema: <http://schema.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
CONSTRUCT {
  ?resource8 schema:containedInPlace ?resource1.
  ?resource8 schema:identifier ?resource11.
  ?resource8 schema:name ?resource12.
  ?resource8 schema:description ?resource13.
  ?resource8 schema:position ?resource14.
  ?resource8 schema:alternateName ?resource15.
  ?resource8 rdf:type ?resource10.
}
WHERE {
  VALUES ?resource1 {
    <https://ld.admin.ch/country/CHE>
  }
  ?resource1 ^schema:containedInPlace ?resource8.
  ?resource8 rdf:type ?resource10.
  VALUES ?resource10 {
    <https://environment.ld.admin.ch/foen/nfi/UnitOfReference/366>
  }
  { ?resource8 schema:identifier ?resource11. }
  UNION
  { ?resource8 schema:name ?resource12. }
  UNION
  { ?resource8 schema:description ?resource13. }
  UNION
  { ?resource8 schema:position ?resource14. }
  UNION
  { ?resource8 schema:alternateName ?resource15. }
}"
`;

exports[`UselessUnionRemover subquery.rq 1`] = `
"PREFIX schema: <http://schema.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
CONSTRUCT {
  ?resource8 schema:containedInPlace ?resource1.
  ?resource8 schema:identifier ?resource11.
  ?resource8 schema:name ?resource12.
  ?resource8 schema:description ?resource13.
  ?resource8 schema:position ?resource14.
  ?resource8 schema:alternateName ?resource15.
  ?resource8 rdf:type ?resource10.
}
WHERE {
  {
    {
      SELECT ?resource8 ?resource11 ?resource12 ?resource13 ?resource14 ?resource15 ?resource10 WHERE {
        VALUES ?resource1 {
          <https://ld.admin.ch/country/CHE>
        }
        ?resource1 ^schema:containedInPlace ?resource8.
        ?resource8 rdf:type ?resource10.
        VALUES ?resource10 {
          <https://environment.ld.admin.ch/foen/nfi/UnitOfReference/366>
        }
        { ?resource8 schema:identifier ?resource11. }
        UNION
        { ?resource8 schema:name ?resource12. }
        UNION
        { ?resource8 schema:description ?resource13. }
        UNION
        { ?resource8 schema:position ?resource14. }
        UNION
        { ?resource8 schema:alternateName ?resource15. }
      }
    }
  }
}"
`;

exports[`UselessUnionRemover union-becomes-optional.rq 1`] = `
"PREFIX schema: <http://schema.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
CONSTRUCT {
  ?resource8 schema:containedInPlace ?resource1.
  ?resource8 schema:identifier ?resource11.
  ?resource8 rdf:type ?resource10.
}
WHERE {
  VALUES ?resource1 {
    <https://ld.admin.ch/country/CHE>
  }
  ?resource1 ^schema:containedInPlace ?resource8.
  ?resource8 rdf:type ?resource10.
  VALUES ?resource10 {
    <https://environment.ld.admin.ch/foen/nfi/UnitOfReference/366>
  }
  OPTIONAL { ?resource8 schema:identifier ?resource11. }
}"
`;

exports[`UselessUnionRemover values.rq 1`] = `
"PREFIX schema: <http://schema.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
SELECT * WHERE {
  VALUES ?resource1 {
    <https://ld.admin.ch/country/CHE>
  }
  ?resource1 ^schema:containedInPlace ?resource8.
  ?resource8 rdf:type ?resource10.
  VALUES ?resource10 {
    <https://environment.ld.admin.ch/foen/nfi/UnitOfReference/366>
  }
  { ?resource8 schema:identifier ?resource11. }
  UNION
  { ?resource8 schema:name ?resource12. }
  UNION
  { ?resource8 schema:description ?resource13. }
  UNION
  { ?resource8 schema:position ?resource14. }
  UNION
  { ?resource8 schema:alternateName ?resource15. }
}"
`;

exports[`UselessUnionRemover values-in-subquery.rq 1`] = `
"PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX hydra: <http://www.w3.org/ns/hydra/core#>
CONSTRUCT {
  ?resource1 hydra:member ?resource2.
  ?resource1 hydra:totalItems ?resource7.
}
WHERE {
  VALUES ?resource1 {
    <http://example.com/people>
  }
  {
    ?resource1 ((rdf:type*)/hydra:memberAssertion) ?resource4.
    ?resource4 hydra:property ?resource6.
    VALUES ?resource6 {
      rdf:type
    }
    ?resource4 hydra:object ?resource3.
    ?resource3 ^rdf:type ?resource2.
  }
  UNION
  {
    SELECT (COUNT(?resource8) AS ?resource7) WHERE {
      ?resource1 ((rdf:type*)/hydra:memberAssertion) ?resource10.
      ?resource10 hydra:property ?resource12.
      VALUES ?resource12 {
        rdf:type
      }
      ?resource10 hydra:object ?resource9.
      ?resource9 ^rdf:type ?resource8.
      VALUES ?resource1 {
        <http://example.com/people>
      }
    }
  }
}"
`;

exports[`UselessUnionRemover values-unique.rq 1`] = `
"PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX schema: <http://schema.org/>
CONSTRUCT {
  ?resource1 rdf:type schema:Person.
  ?resource1 <http://example.org/motherName> ?resource2.
  ?resource1 <http://example.org/fatherName> ?resource6.
}
WHERE {
  ?resource1 rdf:type schema:Person.
  {
    ?resource1 schema:parent ?resource3.
    ?resource3 schema:gender ?resource5.
    VALUES ?resource5 {
      schema:Female
    }
    ?resource3 schema:givenName ?resource2.
  }
  UNION
  {
    ?resource1 schema:parent ?resource7.
    ?resource7 schema:gender ?resource9.
    VALUES ?resource9 {
      schema:Male
    }
    ?resource7 schema:givenName ?resource6.
  }
}"
`;
